File: Pensamiento Computacional en Práctica La Lógica Universal del Código a C++, Java, C#, JavaScript y Python/Codigo GO/anexo.go
package main

import (
	"fmt"
	"strconv"
)

var productos [][]string
var ventas [][]string

func main() {
	productos = CargarProductos()
	ventas = CrearVenta()

	MenuPrincipal(productos, ventas)
}

// MenuPrincipal (reestructurado) es el bucle principal de la aplicación.
func MenuPrincipal(vproductos [][]string, vventas [][]string) {
	var opcion int
	var salida bool
	var idticket string
	datosmenuprincipal := []string{"Productos", "Punto de Venta", "Inventario", "Ventas", "Salida"}
	narchivo := Fecha() + ".csv"

	if ExisteArchivo(narchivo) {
		idticket = LeerDatosCSV(narchivo)
	}

	for !salida {
		if idticket == "" {
			idticket = ObtenerUltimoValorVentas(vventas)
		}
		opcionStr := DesplegarMenu("Menu de Punto de Tienda de Abarrotes la Pequeña", datosmenuprincipal)

		if opcionStr == "" {
			fmt.Println("opcion incorrecta")
		} else {
			opcion, _ = strconv.Atoi(opcionStr)
			switch opcion {
			case 1:
				MenuProductos(vproductos)
			case 2:
				MenuPuntoVenta(vventas, idticket, vproductos, narchivo)
				idticket = ""
			case 3:
				MenuInventario(vproductos)
			case 4:
				fmt.Println(MostrarListaVentas(vventas))
			case 5:
				fmt.Println("Salida del Sistema")
				salida = true
			default:
				fmt.Println("No existe esta opcion")
			}
		}
	}
}

// MenuPuntoVenta (actualizado) gestiona el menú del punto de venta para un ticket específico.
func MenuPuntoVenta(vventas [][]string, idticket string, vproductos [][]string, narchivo string) {
	var opcion, membrete string
	pago := false
	tamticket := 50
	Vticket := make([][]string, tamticket)
	idticket = IdTicketSiguiente(idticket)
	fechadia := Fecha()

	do {
		membrete = "Fecha del Día: " + fechadia + " | Ticket No: " + idticket + "
"
		membrete += "-----------------------------------------------------
"
		Tickettexto := MostrarTicket(Vticket)
		if Tickettexto != "" {
			membrete += "
" + Tickettexto + "
"
		}
		datosmenu := []string{"Agregar", "Eliminar", "Listado", "Pagar", "Salida"}
		opcion = DesplegarMenu(membrete+"
Menú de Punto de Venta", datosmenu)

		if opcion == "" {
			fmt.Println("Dato incorrecto introducido")
		} else {
			op, _ := strconv.Atoi(opcion)
			switch op {
			case 1:
				CapturaVentaProducto(Vticket, vproductos, idticket, tamticket)
			case 2:
				Eliminar(Vticket, vproductos)
			case 3:
				if MostrarTicket(Vticket) != "" {
					fmt.Println(MostrarTicket(Vticket))
				}
			case 4:
				fmt.Println(MostrarTicket(Vticket))
				Pagar(idticket, vventas, Vticket)
				if ObtenerUltimaPosicion(Vticket) > -1 {
					fmt.Println("tamaño de venta creando archivo", ObtenerUltimaPosicion(vventas), ObtenerUltimaPosicion(Vticket))
					GuardarDatosCsvVentaTicket(Vticket, narchivo, ObtenerUltimaPosicion(Vticket)+1, idticket)
					subtotal := fmt.Sprintf("%.2f", SubTotalTicket(Vticket))
					iva := fmt.Sprintf("%.2f", IvaTicket(Vticket))
					total := fmt.Sprintf("%.2f", TotalTicket(Vticket))
					cadena := GenerarCSVFilaVenta(fechadia, idticket, subtotal, iva, total)
					GuardarDatosCsvVentas("ventas.csv", cadena)
				}
				pago = true
				opcion = "5"
			case 5:
				fmt.Println("Salida del sistema")
				if !pago {
					fmt.Println("No pagó el ticket. Ejecutando devolución...")
					DevolucionTicket(Vticket, vproductos)
					fmt.Println("Eliminando ticket", idticket)
				}
			default:
				fmt.Println("Opción no válida")
			}
		}
	} while opcion != "5"
}

File: Pensamiento Computacional en Práctica La Lógica Universal del Código a C++, Java, C#, JavaScript y Python/Codigo RUST/anexo.rs
use std::io;

mod algoritmos_anexo;
mod algoritmos_capitulo3;
mod algoritmos_capitulo4;
mod algoritmos_capitulo5;
mod algoritmos_capitulo6;
mod algoritmos_capitulo7;

fn main() {
    let mut productos = cargar_productos();
    let mut ventas = crear_venta();

    menu_principal(&mut productos, &mut ventas);
}

fn menu_principal(vproductos: &mut [[String; 4]; 10], vventas: &mut Vec<[String; 5]>) {
    let mut opcion: i32;
    let mut salida = false;
    let mut idticket = "".to_string();
    let datosmenuprincipal = vec!["Productos", "Punto de Venta", "Inventario", "Ventas", "Salida"];
    let narchivo = format!("{}.csv", fecha());

    if existe_archivo(&narchivo) {
        if let Ok(id) = leer_datos_csv(&narchivo) {
            idticket = id;
        }
    }

    while !salida {
        if idticket.is_empty() {
            idticket = obtener_ultimo_valor_ventas(vventas);
        }
        let opcion_str = desplegar_menu("Menu de Punto de Tienda de Abarrotes la Pequeña", &datosmenuprincipal);

        if opcion_str.is_empty() {
            println!("opcion incorrecta");
        } else {
            opcion = opcion_str.parse().unwrap_or(0);
            match opcion {
                1 => menu_productos(vproductos),
                2 => {
                    menu_punto_venta(vventas, &mut idticket, vproductos, &narchivo);
                    idticket = "".to_string();
                }
                3 => menu_inventario(vproductos),
                4 => println!("{}", mostrar_lista_ventas(vventas)),
                5 => {
                    println!("Salida del Sistema");
                    salida = true;
                }
                _ => println!("No existe esta opcion"),
            }
        }
    }
}

fn menu_punto_venta(
    vventas: &mut Vec<[String; 5]>,
    idticket: &mut String,
    vproductos: &mut [[String; 4]; 10],
    narchivo: &str,
) {
    let mut opcion: String;
    let mut pago = false;
    const TAMTICKET: usize = 50;
    let mut vticket = Vec::with_capacity(TAMTICKET);
    *idticket = id_ticket_siguiente(idticket);
    let fechadia = fecha();

    loop {
        let mut membrete = format!("Fecha del Día: {} | Ticket No: \n", fechadia, idticket);
        membrete.push_str("-----------------------------------------------------\
");
        let ticket_texto = mostrar_ticket(&vticket);
        if !ticket_texto.is_empty() {
            membrete.push_str(&format!("\n{}\n", ticket_texto));
        }
        let datosmenu = vec!["Agregar", "Eliminar", "Listado", "Pagar", "Salida"];
        opcion = desplegar_menu(&format!("{}\nMenú de Punto de Venta", membrete), &datosmenu);

        if opcion.is_empty() {
            println!("Dato incorrecto introducido");
        } else {
            let op: i32 = opcion.parse().unwrap_or(0);
            match op {
                1 => captura_venta_producto(&mut vticket, vproductos, idticket, TAMTICKET),
                2 => eliminar(&mut vticket, vproductos),
                3 => {
                    if !mostrar_ticket(&vticket).is_empty() {
                        println!("{}", mostrar_ticket(&vticket));
                    }
                }
                4 => {
                    println!("{}", mostrar_ticket(&vticket));
                    pagar(idticket, vventas, &vticket);
                    if !vticket.is_empty() {
                        println!(
                            "tamaño de venta creando archivo {} {}",
                            vventas.len(),
                            vticket.len()
                        );
                        if let Err(e) = guardar_datos_csv_venta_ticket(&vticket, narchivo, idticket) {
                            println!("Error al guardar el ticket: {}", e);
                        }
                        let subtotal = format!("{{:.2}}", sub_total_ticket(&vticket));
                        let iva = format!("{{:.2}}", iva_ticket(&vticket));
                        let total = format!("{{:.2}}", total_ticket(&vticket));
                        let cadena = generar_csv_fila_venta(&fechadia, idticket, &subtotal, &iva, &total);
                        if let Err(e) = guardar_datos_csv_ventas("ventas.csv", &cadena) {
                            println!("Error al guardar la venta: {}", e);
                        }
                    }
                    pago = true;
                    break;
                }
                5 => {
                    println!("Salida del sistema");
                    if !pago {
                        println!("No pagó el ticket. Ejecutando devolución...");
                        devolucion_ticket(&vticket, vproductos);
                        println!("Eliminando ticket {}", idticket);
                    }
                    break;
                }
                _ => println!("Opción no válida"),
            }
        }
    }
}
